runtime/compiler/trj9/control/CompilationThread.cpp

```diff
@@ -6403,97 +6377,41 @@ TR::CompilationInfoPerThreadBase::compile(J9VMThread * vmThread,
-            {
-            // NOTE:
-            //       for intentional crashes, intentional traps cause the dump, and we
-            //       are not protected against them (flag: J9PORT_SIG_FLAG_SIGTRAP)
-            protectedResult = j9sig_protect((j9sig_protected_fn)wrappedCompile, static_cast<void *>(&compParam),
-                                               (j9sig_handler_fn)  blankDumpSignalHandler, vmThread,
-                                               flags, &result);
-            }
-         else
-            {
-            protectedResult = j9sig_protect((j9sig_protected_fn)wrappedCompile, static_cast<void*>(&compParam),
-                                               (j9sig_handler_fn)  jitSignalHandler, vmThread,
-                                               flags, &result);
-            }
-
-         if (protectedResult == 0) // successful completion of wrappedCompile
-            metaData = reinterpret_cast<TR_MethodMetaData *>(result);
-         else
-            metaData = 0;
-         }
-      else
-#endif
-         metaData = wrappedCompile(privatePortLibrary, &compParam);
-      }
-   else
-      {
-      entry->_compErrCode = compilationRestrictedMethod;
-      // can't compile this method: mark it appropriately
-      //if (!entry->isDLTCompile())
-      //   startPC = compilationEnd(vmThread, jitConfig, method,
-      //                            0, entry->_oldStartPC, entry->_classForNewInstance, vm);
-      }
-
-   if (vmThread)
-      {
-      vmThread->omrVMThread->vmState = oldState;
-      }
-
-
+   return vm;
+   }
+
+/**
+ * @brief TR::CompilationInfoPerThreadBase::postCompilationTasks
+ * @param vmThread Pointer to the current J9VMThread
+ * @param entry Pointer to the method to be compiled entry
+ * @param method Pointer to the J9Method
+ * @param aotCachedMethod Pointer to the AOT code to be relocated (if it exists)
+ * @param metaData Pointer to the TR_MethodMetaData created on a successful compile
+ * @param compParam Reference to the CompileParameters associated with the current compilation
+ * @param canDoRelocatableCompile Indicates if the current compilation could be AOT compiled
+ * @param eligibleForRelocatableCompile Indicates if the current compilation was eligible for AOT
+ * @param reloRuntime Pointer to a TR_RelocationRuntime; NULL if JVM does not have AOT support
+ * @param scratchMemoryAvailable Indicates if the scratch memory infrastructre for the current compilation can be used
+ * @return StartPC of the compiled body, NULL if the compilation failed
+ *
+ * This method is used to perform a set of tasks needed after attempting a compilation. It is important to note
+ * that the scratch memory infrastructure (TR_Memory, TR::Region, TR::SegmentAllocator) may not exist within this
+ * method as this method is called in both the try and catch block in TR::CompilationInfoPerThreadBase::compile.
+ */
+void *
+TR::CompilationInfoPerThreadBase::postCompilationTasks(J9VMThread * vmThread,
```

Comment:
Can this method throw? If so at what line? Since this method is executed in the catch block as well it's better if we can ensure that it can never throw.

Reply:
I don't think anything explicitly throws right now. Any exception that could be thrown would likely be a `std::bad_alloc()` thrown because of some memory allocated via the `TR_Memory` or `TR::Region` associated with the `comp` object; however, that would require the `comp` object to exist which could only happen in the try block (since it is set to NULL in the catch block). The other possibility is if `acquireVMaccessIfNeeded` is called, which can throw `TR::CompilationInterrupted` or `J9::JITShutdown`, though I don't think that's called at any point in `postCompilationTasks`. 

So I suppose it's technically possible for something to throw, but no one should since there is a problem with a throw within `postCompilationTasks`, namely:
```
   // Re-acquire the compilation monitor so that we can update the J9Method.
   //
   _compInfo.debugPrint(vmThread, "\tacquiring compilation monitor\n");
   _compInfo.acquireCompMonitor(vmThread);
   _compInfo.debugPrint(vmThread, "+CM\n");
   if (_onSeparateThread)
      {
      // Acquire the queue slot monitor now
      //
      _compInfo.debugPrint(vmThread, "\tre-acquiring queue-slot monitor\n");
      entry->acquireSlotMonitor(vmThread);
      _compInfo.debugPrint(vmThread, "+AM-", entry);
      }
```
which could end up running twice.

I'll update the comment at the top to state that an exception should not escape `post CompilationTasks`.